File: internal/trace/trace_test.go
package trace

import (
	"bufio"
	"context"
	"encoding/json"
	"os"
	"strings"
	"sync"
	"testing"
)

func TestBusWritesOrderedParseableJSONL(t *testing.T) {
	directory := t.TempDir()
	sink, err := OpenFileSink(directory, "ordered-session")
	if err != nil {
		t.Fatal(err)
	}
	path := sink.Path()
	bus, err := NewBus("ordered-session", sink)
	if err != nil {
		t.Fatal(err)
	}

	const events = 24
	var wait sync.WaitGroup
	wait.Add(events)
	for index := range events {
		go func() {
			defer wait.Done()
			if err := bus.Emit(context.Background(), "probe", map[string]int{"index": index}); err != nil {
				t.Errorf("emit: %v", err)
			}
		}()
	}
	wait.Wait()
	if err := bus.Close(); err != nil {
		t.Fatal(err)
	}

	file, err := os.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	count := 0
	for scanner.Scan() {
		count++
		var event Event
		if err := json.Unmarshal(scanner.Bytes(), &event); err != nil {
			t.Fatalf("line %d is not valid JSON: %v", count, err)
		}
		if event.Version != SchemaVersion {
			t.Fatalf("line %d version = %d, want %d", count, event.Version, SchemaVersion)
		}
		if event.SessionID != "ordered-session" {
			t.Fatalf("line %d session = %q", count, event.SessionID)
		}
		if event.Sequence != uint64(count) {
			t.Fatalf("line %d sequence = %d", count, event.Sequence)
		}
		if event.Type != "probe" || len(event.Data) == 0 || event.Time.IsZero() {
			t.Fatalf("line %d has incomplete event: %#v", count, event)
		}
	}
	if err := scanner.Err(); err != nil {
		t.Fatal(err)
	}
	if count != events {
		t.Fatalf("event count = %d, want %d", count, events)
	}

	info, err := os.Stat(path)
	if err != nil {
		t.Fatal(err)
	}
	if got := info.Mode().Perm(); got != 0o600 {
		t.Fatalf("trace permissions = %o, want 600", got)
	}
}

func TestOpenFileSinkRejectsInvalidOrDuplicateSession(t *testing.T) {
	directory := t.TempDir()
	invalid := []string{"", "../escape", "has/slash", "-leading", "has space", strings.Repeat("a", 129)}
	for _, sessionID := range invalid {
		t.Run("invalid_"+sessionID, func(t *testing.T) {
			if sink, err := OpenFileSink(directory, sessionID); err == nil {
				sink.Close()
				t.Fatalf("expected invalid session ID %q to fail", sessionID)
			}
		})
	}

	sink, err := OpenFileSink(directory, "duplicate")
	if err != nil {
		t.Fatal(err)
	}
	if err := sink.Close(); err != nil {
		t.Fatal(err)
	}
	if duplicate, err := OpenFileSink(directory, "duplicate"); err == nil {
		duplicate.Close()
		t.Fatal("expected reopening the same session to fail")
	}
}

func TestNewSessionIDProducesFileSafeID(t *testing.T) {
	first, err := NewSessionID()
	if err != nil {
		t.Fatal(err)
	}
	second, err := NewSessionID()
	if err != nil {
		t.Fatal(err)
	}
	if first == second {
		t.Fatalf("session IDs unexpectedly match: %q", first)
	}
	if !sessionIDPattern.MatchString(first) || !sessionIDPattern.MatchString(second) {
		t.Fatalf("session IDs are not file safe: %q, %q", first, second)
	}
}

File: src/trace/trace.test.ts
import assert from "node:assert/strict";
import { mkdtemp, readFile, rm, stat } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import test from "node:test";

import {
  EventBus,
  type EventSink,
  TRACE_SCHEMA_VERSION,
  newSessionId,
  type TraceEvent,
} from "./events.js";
import { FileEventSink } from "./file-event-sink.js";

async function traceDirectory(t: test.TestContext): Promise<string> {
  const directory = await mkdtemp(path.join(os.tmpdir(), "nash-trace-"));
  t.after(async () => {
    await rm(directory, { recursive: true, force: true });
  });
  return directory;
}

test("EventBus serializes concurrent events as ordered parseable JSONL", async (t) => {
  const directory = await traceDirectory(t);
  const sink = await FileEventSink.open(directory, "ordered-session");
  const bus = new EventBus("ordered-session", [sink]);

  await Promise.all(
    Array.from({ length: 24 }, (_, index) =>
      bus.emit("probe", { index }),
    ),
  );
  await bus.close();

  const lines = (await readFile(sink.path, "utf8")).trimEnd().split("\n");
  assert.equal(lines.length, 24);
  const events = lines.map((line) => JSON.parse(line) as TraceEvent);
  for (const [index, event] of events.entries()) {
    assert.equal(event.version, TRACE_SCHEMA_VERSION);
    assert.equal(event.sessionId, "ordered-session");
    assert.equal(event.sequence, index + 1);
    assert.equal(event.type, "probe");
    assert.equal((event.data as { readonly index: number }).index, index);
    assert.ok(!Number.isNaN(Date.parse(event.time)));
  }

  assert.equal((await stat(sink.path)).mode & 0o777, 0o600);
});

test("EventBus snapshots data when emit is called", async (t) => {
  const directory = await traceDirectory(t);
  const sink = await FileEventSink.open(directory, "snapshot-session");
  const bus = new EventBus("snapshot-session", [sink]);
  const data = { nested: { value: "before" } };

  const emitted = bus.emit("snapshot", data);
  data.nested.value = "after";
  await emitted;
  await bus.close();

  const event = JSON.parse(
    (await readFile(sink.path, "utf8")).trim(),
  ) as TraceEvent;
  assert.deepEqual(event.data, { nested: { value: "before" } });
});

test("EventBus keeps sink failures sticky and closes every sink", async () => {
  const writeFailure = new Error("sink write failed");
  const failing = new FailingSink(writeFailure);
  const other = new TrackingSink();
  const bus = new EventBus("failing-session", [failing, other]);

  await assert.rejects(bus.emit("first", { index: 0 }), writeFailure);
  await assert.rejects(bus.emit("second", { index: 1 }), writeFailure);
  await assert.rejects(
    bus.close(),
    (error: unknown) =>
      error instanceof AggregateError && error.errors.includes(writeFailure),
  );

  assert.equal(failing.closed, true);
  assert.equal(other.closed, true);
  assert.equal(failing.writes, 1);
  assert.equal(other.writes, 0);
});

test("FileEventSink refuses to clobber an existing session trace", async (t) => {
  const directory = await traceDirectory(t);
  const first = await FileEventSink.open(directory, "duplicate");
  await first.close();

  await assert.rejects(FileEventSink.open(directory, "duplicate"));
});

test("FileEventSink rejects unsafe session identifiers", async (t) => {
  const directory = await traceDirectory(t);
  const invalid = [
    "",
    "../escape",
    "has/slash",
    "-leading",
    "has space",
    "a".repeat(129),
  ];

  for (const sessionId of invalid) {
    await assert.rejects(FileEventSink.open(directory, sessionId));
  }
});

test("newSessionId returns distinct file-safe identifiers", () => {
  const first = newSessionId(new Date("2026-08-30T12:34:56.789Z"));
  const second = newSessionId(new Date("2026-08-30T12:34:56.789Z"));

  assert.match(first, /^20260830T123456Z-[0-9a-f]{8}$/);
  assert.match(second, /^20260830T123456Z-[0-9a-f]{8}$/);
  assert.notEqual(first, second);
});

class FailingSink implements EventSink {
  public writes = 0;
  public closed = false;

  public constructor(private readonly failure: Error) {}

  public async write(_event: TraceEvent): Promise<void> {
    this.writes += 1;
    throw this.failure;
  }

  public async close(): Promise<void> {
    this.closed = true;
  }
}

class TrackingSink implements EventSink {
  public writes = 0;
  public closed = false;

  public async write(_event: TraceEvent): Promise<void> {
    this.writes += 1;
  }

  public async close(): Promise<void> {
    this.closed = true;
  }
}
